core/field_registry.ts

```diff
@@ -68,7 +68,7 @@ export function fromJson(options: AnyDuringMigration): Field|null {
  */
 function fromJsonInternal(options: AnyDuringMigration): Field|null {
   const fieldObject =
-      registry.getObject(registry.Type.FIELD, options['type']) as
+      registry.getObject(registry.Type.FIELD, options['type']) as unknown as
```

Comment:
This smells bad.  From your bug report it looks like `tsc` is correctly deducing that `getObject` is returning a `Field`, and the error message (that `Field` is not assignable to `IRegisterableField` appears to be legit.

I think part of the problem may be that `IRegisterableField` is now incorrectly defined: [it was originally specifically _not_ an `@interface`](https://github.com/google/blockly/blob/d86cec9145edd36d7d0f89806371bc184116c948/core/interfaces/i_registrable_field.js#L27) because it's really supposed to be declaring a type that has a certain `static` method (i.e., property on the constructor function object) named `fromJson`, rather than declaring that it is a type with an _instance_ method (i.e., property on the `.prototype` object) named `from Json`, but this appears to have been erroneously converted to an `interface` by PR #6299.

So I think the fix here is one or more of the following:

1. Correct [the definition of `IRegisterableField`](https://github.com/google/blockly/blob/883d78d5a09a6f45b4982943e57c5e9dbd3fbc7e/core/interfaces/i_registrable_field.ts#L29), possibly by using [this decorator pattern](https://stackoverflow.com/a/43674389/4969945).
2. Correct [the definition of of `Type.FIELD`](https://github.com/google/blockly/blob/883d78d5a09a6f45b4982943e57c5e9dbd3fbc7e/core/registry.ts#L81) to use `IRegisterableField` instead of `Field`.
3. Add a runtime typecheck to make sure the `Field` returned by `registry.getObject` is actually an `IRegisterableField`—tricky, since interfaces are not concrete types that can be checked with `instanceof`, but possibly something like this might work:
   ```
   if (typeof Object.getPrototypeOf(fieldObject).constructor.fromJson !== 'function) {
     throw new TypeError('returned Field was not a IRegisterableField');
    }
   ```
   (I'm not sure `tsc` is clever enough to do the necessary inference, and in any case this only checks for the existence of a `fromJson` static method, not that method's signature.)

Maybe @maribethb can weigh in on what the correct type of the template parameter to `registry.Type.FIELD` should be; I'm weirded out by the lack of overlap between `Field` and `IRegisterableField` which seems like it ought to be a subclass of the former.

Reply:
Agreed that this is not ideal. I don't know enough about the code base to make a decision though. Will wait for suggestion.

Reply:
@maribethb, do you have an opinion on this?

Reply:
I think the best/simplest option for now is Option 3. Do a runtime check that the fieldObject (which is actually a field constructor) has a fromJson method on it, and throw a useful error if it does not. That way we know we at least have runtime safety.

We can discuss better solutions to defining the static method on a type later.